vpr/src/pack/cluster_legalizer.cpp

```diff
@@ -997,11 +997,70 @@ static void update_molecule_chain_info(t_pack_molecule* chain_molecule, const t_
     VTR_ASSERT(false);
 }
 
+/**
+ * @brief Revalidate the molecules associated with this pb. The mol_validated
+ *        flag is used to check if the mol has already been validated.
+ */
+static void revalid_molecules(const t_pb* pb,
+                              bool &mol_validated,
+                              const Prepacker& prepacker) {
+    const t_pb_type* pb_type = pb->pb_graph_node->pb_type;
+
+    if (pb_type->blif_model == nullptr) {
+        int mode = pb->mode;
+        for (int i = 0; i < pb_type->modes[mode].num_pb_type_children && pb->child_pbs != nullptr; i++) {
+            for (int j = 0; j < pb_type->modes[mode].pb_type_children[i].num_pb && pb->child_pbs[i] != nullptr; j++) {
+                if (pb->child_pbs[i][j].name != nullptr || pb->child_pbs[i][j].child_pbs != nullptr) {
+                    revalid_molecules(&pb->child_pbs[i][j], mol_validated, prepacker);
+                }
+            }
+        }
+    } else {
+        //Primitive
+        auto& atom_ctx = g_vpr_ctx.mutable_atom();
+
+        auto blk_id = atom_ctx.lookup.pb_atom(pb);
+        if (blk_id) {
+            /* If any molecules were marked invalid because of this logic block getting packed, mark them valid */
+
+            //Update atom netlist mapping
+            atom_ctx.lookup.set_atom_clb(blk_id, ClusterBlockId::INVALID());
+            atom_ctx.lookup.set_atom_pb(blk_id, nullptr);
+
+            t_pack_molecule* cur_molecule = prepacker.get_atom_molecule(blk_id);
+            if (!mol_validated) {
+                int i;
+                for (i = 0; i < get_array_size_of_molecule(cur_molecule); i++) {
+                    if (cur_molecule->atom_block_ids[i]) {
+                        if (atom_ctx.lookup.atom_clb(cur_molecule->atom_block_ids[i]) != ClusterBlockId::INVALID()) {
+                            break;
+                        }
+                    }
+                }
+                /* All atom blocks are open for this molecule, place back in queue */
+                if (i == get_array_size_of_molecule(cur_molecule)) {
+                    mol_validated = true;
+                    // when invalidating a molecule check if it's a chain molecule
+                    // that is part of a long chain. If so, check if this molecule
+                    // have modified the chain_id value based on the stale packing
+                    // then reset the chain id and the first packed molecule pointer
+                    // this is packing is being reset
+                    if (cur_molecule->is_chain() && cur_molecule->chain_info->is_long_chain && cur_molecule->chain_info->first_packed_molecule == cur_molecule) {
+                        cur_molecule->chain_info->first_packed_molecule = nullptr;
+                        cur_molecule->chain_info->chain_id = -1;
+                    }
+                }
+            }
+        }
+    }
+}
+
 /*
  * @brief Revert trial atom block iblock and free up memory space accordingly.
  */
```

Comment:
Good to expand this comment with as much as you know; you should at least know what mol_validated does so it could be commented.

Reply:
See previous response. I agree that the mol_validated variable was confusing and reworked the function to make it better.